File: core-py/src/openglass_core/attest/identity.py
"""Filesystem-backed Ed25519 identity storage — the Python counterpart to
core-js's `attest/identity.ts`. Not a secrets vault; see that file's docstring for the
same caveat.
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import TypedDict

from openglass import AgentIdentity, base64url_decode, base64url_encode, generate_ed25519_keypair


class _StoredIdentity(TypedDict, total=False):
    agentId: str
    kid: str
    privateKey: str
    publicKey: str


def load_or_create_identity(path: str | Path) -> AgentIdentity:
    p = Path(path)
    if p.exists():
        return read_identity(p)
    keypair = generate_ed25519_keypair()
    identity = AgentIdentity(kid="new", private_key=keypair.private_key, public_key=keypair.public_key)
    persist_identity(p, identity)
    return identity


def read_identity(path: str | Path) -> AgentIdentity:
    stored: _StoredIdentity = json.loads(Path(path).read_text())
    return AgentIdentity(
        agent_id=stored.get("agentId"),
        kid=stored["kid"],
        private_key=base64url_decode(stored["privateKey"]),
        public_key=base64url_decode(stored["publicKey"]),
    )


def persist_identity(path: str | Path, identity: AgentIdentity) -> None:
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    stored: _StoredIdentity = {
        "kid": identity.kid,
        "privateKey": base64url_encode(identity.private_key),
        "publicKey": base64url_encode(identity.public_key),
    }
    if identity.agent_id is not None:
        stored["agentId"] = identity.agent_id
    p.write_text(json.dumps(stored, indent=2))
    os.chmod(p, 0o600)

File: core-js/src/attest/identity.ts
import { mkdirSync, readFileSync, writeFileSync, existsSync } from "node:fs";
import { dirname } from "node:path";
import { base64UrlDecode, base64UrlEncode, generateEd25519KeyPair, type AgentIdentity } from "openglass-sdk";

interface StoredIdentity {
  agentId?: string;
  kid: string;
  privateKey: string;
  publicKey: string;
}

/**
 * Loads an Ed25519 identity from a local JSON file, generating and persisting a fresh
 * one on first use. This is "key storage" in the sense SPEC D13 means for a local
 * integration: the private key lives in this process (and this one file), never on
 * OpenGlass's servers and never over the network except as a signature.
 *
 * Not a secrets vault — if your deployment needs one (KMS, Vault, etc.), read the key
 * from there yourself and construct an `AgentIdentity` directly; this helper is the
 * zero-config default for the common case of one process on one machine.
 */
export function loadOrCreateIdentity(path: string): AgentIdentity {
  if (existsSync(path)) return readIdentity(path);
  const { privateKey, publicKey } = generateEd25519KeyPair();
  const identity: AgentIdentity = { kid: "new", privateKey, publicKey };
  persistIdentity(path, identity);
  return identity;
}

export function readIdentity(path: string): AgentIdentity {
  const stored = JSON.parse(readFileSync(path, "utf8")) as StoredIdentity;
  return {
    agentId: stored.agentId,
    kid: stored.kid,
    // base64UrlDecode returns a Node Buffer; normalize to a plain Uint8Array so callers
    // (and tests) see the same concrete type generateEd25519KeyPair() produces.
    privateKey: new Uint8Array(base64UrlDecode(stored.privateKey)),
    publicKey: new Uint8Array(base64UrlDecode(stored.publicKey)),
  };
}

/** Call this again after registration to persist the server-assigned `agentId`/`kid`. */
export function persistIdentity(path: string, identity: AgentIdentity): void {
  mkdirSync(dirname(path), { recursive: true });
  const stored: StoredIdentity = {
    agentId: identity.agentId,
    kid: identity.kid,
    privateKey: base64UrlEncode(identity.privateKey),
    publicKey: base64UrlEncode(identity.publicKey),
  };
  writeFileSync(path, JSON.stringify(stored, null, 2), { mode: 0o600 });
}
